### PR TITLE
8327621: Check return value of uname in os::get_host_name

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

Trivial bug fix where we didn't check the return value of `uname(2)` on POSIX platforms. I checked callers of `get_host_name` and all wrap the calls in an `if` branch so they already handle failures.

Regards,
Johan

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327621](https://bugs.openjdk.org/browse/JDK-8327621): Check return value of uname in os::get_host_name (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [9094ac42](https://git.openjdk.org/jdk/pull/18157/files/9094ac42cb6824708c620a64e7b97b14c3783480)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18157/head:pull/18157` \
`$ git checkout pull/18157`

Update a local copy of the PR: \
`$ git checkout pull/18157` \
`$ git pull https://git.openjdk.org/jdk.git pull/18157/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18157`

View PR using the GUI difftool: \
`$ git pr show -t 18157`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18157.diff">https://git.openjdk.org/jdk/pull/18157.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18157#issuecomment-1984177532)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
Error checking in general is a good thing  but in this case there are no specified failure modes for uname under Posix, and for Linux the only potential error is
```
 EFAULT buf is not valid
```
which is obviously not a concern in the code we have.

Did you encounter an error or was this just defensive programming?

**Comment by jdksjolen** [comment]
> Error checking in general is a good thing but in this case there are no specified failure modes for uname under Posix, and for Linux the only potential error is
> 
> ```
>  EFAULT buf is not valid
> ```
> 
> which is obviously not a concern in the code we have.
> 
> Did you encounter an error or was this just defensive programming?

This is simply defensive programming. I don't know what the potential issues could be on OSX or Solaris.



> _Mailing list message from [Peter Tribble](mailto:peter.tribble@gmail.com) on [hotspot-runtime-dev](mailto:hotspot-runtime-dev@mail.openjdk.org):_
> 
> On Thu, Mar 7, 2024 at 6:29?PM Johan Sj?len <jsjolen at openjdk.org> wrote:
> 
> > Hi,
> > Trivial bug fix where we didn't check the return value of `uname(2)` on
> > POSIX platforms. I checked callers of `get_host_name` and all wrap the
> > calls in an `if` branch so they already handle failures.
> 
> I notice you're checking for a return value of zero. Posix 1003.1 states something like
> 
> Upon successful completion, a non-negative value is returned. Otherwise, ?1 is returned and errno is set to indicate the error.
> 
> So the correct check based on that would be != -1.
> 
> Linux and FreeBSD document returning 0 on success, but Solaris/illumos definitely returns 1 (and as the port maintainer there it would be nice to have this correct). I don't have an AIX box right now but the existing uname(2) check within openjdk is for -1, eg https://github.com/openjdk/jdk/blob/master/src/hotspot/os/aix/os_aix.cpp#L2799
> 
> Thanks,
> 
> -- -Peter Tribble http://www.petertribble.co.uk/ - http://ptribble.blogspot.com/ -------------- next part -------------- An HTML attachment was scrubbed... URL: <https://mail.openjdk.org/pipermail/hotspot-runtime-dev/attachments/20240307/dfa31b6a/attachment.htm>

Thanks for the details Peter, I'll change the comparison as requested.

File: src/hotspot/os/posix/os_posix.cpp [hunk 1]
```diff
@@ -610,9 +610,11 @@ void os::print_jni_name_suffix_on(outputStream* st, int args_size) {
 
 bool os::get_host_name(char* buf, size_t buflen) {
   struct utsname name;
-  uname(&name);
-  jio_snprintf(buf, buflen, "%s", name.nodename);
-  return true;
+  if (uname(&name) == 0) {
+    jio_snprintf(buf, buflen, "%s", name.nodename);
+    return true;
+  }
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/os_posix.cpp:616]: Else I suggest printing a warning and reporting the error using os::strerror

**Review by dholmes-ora — COMMENTED** [review]
Okay on being defensive - one suggestion below.

File: src/hotspot/os/posix/os_posix.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/os/posix/os_posix.cpp:616]: Bikeshedding, I would just report false, but print a warning using UL, flag "os", with warning level. That will be shown by default, but this is strange enough that users should see this.

**Comment by jdksjolen** [comment]
Hi,

My apologies for the slow update on this, I was busy with other things. I changed the check to `!= -1` and performed the UL warning as per David's and Thomas's suggestion.

**Comment by tstuefe** [comment]
Looks good. I wonder though if we should something in case the hostname is truncated on output. Possibly in a separate RFR.

File: src/hotspot/os/posix/os_posix.cpp [hunk 2]
```diff
@@ -610,9 +610,14 @@ void os::print_jni_name_suffix_on(outputStream* st, int args_size) {
...
   struct utsname name;
-  uname(&name);
-  jio_snprintf(buf, buflen, "%s", name.nodename);
-  return true;
+  int retcode = uname(&name);
+  if (retcode != -1) {
+    jio_snprintf(buf, buflen, "%s", name.nodename);
+    return true;
+  }
+  const char* errmsg = os::strerror(retcode);
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/os_posix.cpp:618]: That should be `errno` not `retcode`.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/os/posix/os_posix.cpp (hunk 2 above)
Comment (jdksjolen) [reply src/hotspot/os/posix/os_posix.cpp]: Of course. Thank you for noticing and commenting.

**Review by dholmes-ora — APPROVED** [review]
LGTM thanks

**Comment by jdksjolen** [comment]
/integrate

Thank you, errors seem intermittent and unrelated to this change.

**Pull request closed without merging**